File: go/dsa/algorithms/strings/min-window.go
func minWindow(s string, t string) string {
    if len(s) < len(t) { return "" }
    targetFreq := make(map[byte]int)
    for i := 0; i < len(t); i++ { targetFreq[t[i]]++ }
    
    windowFreq := make(map[byte]int)
    l, r, count, minLen, start := 0, 0, 0, 1<<31-1, 0
    required := len(targetFreq)
    
    for r < len(s) {
        char := s[r]
        windowFreq[char]++
        if targetFreq[char] > 0 && windowFreq[char] == targetFreq[char] {
            count++
        }
        
        for count == required {
            if r-l+1 < minLen {
                minLen = r - l + 1
                start = l
            }
            leftChar := s[l]
            windowFreq[leftChar]--
            if targetFreq[leftChar] > 0 && windowFreq[leftChar] < targetFreq[leftChar] {
                count--
            }
            l++
        }
        r++
    }
    if minLen == 1<<31-1 { return "" }
    return s[start : start+minLen]
}

File: javascript/dsa/algorithms/strings/min-window.js
function minWindow(s, t) {
  if (s.length < t.length) return '';
  const targetFreq = {};
  for (const char of t) targetFreq[char] = (targetFreq[char] || 0) + 1;
  
  const windowFreq = {};
  let l = 0, r = 0, count = 0, minLen = Infinity, startNode = 0;
  const required = Object.keys(targetFreq).length;
  
  while (r < s.length) {
    const char = s[r];
    windowFreq[char] = (windowFreq[char] || 0) + 1;
    if (targetFreq[char] && windowFreq[char] === targetFreq[char]) {
      count++;
    }
    
    while (count === required) {
      if (r - l + 1 < minLen) {
        minLen = r - l + 1;
        startNode = l;
      }
      const leftChar = s[l];
      windowFreq[leftChar]--;
      if (targetFreq[leftChar] && windowFreq[leftChar] < targetFreq[leftChar]) {
        count--;
      }
      l++;
    }
    r++;
  }
  return minLen === Infinity ? '' : s.substring(startNode, startNode + minLen);
}